dotCMS/src/main/java/com/dotmarketing/cms/urlmap/URLMapAPIImpl.java

```diff
@@ -0,0 +1,318 @@
+                        hostField.isRequired() ? "conhost, modDate" : "modDate",
+                        this.wuserAPI.getSystemUser(), true);
+
+        if (!contentletSearches.isEmpty()) {
+            int idx = 0;
+            if (contentletSearches.size() == 2) {
+                // prefer session setting
+                final Contentlet second = this.contentletAPI
+                        .find(contentletSearches.get(1).getInode(),
+                                this.wuserAPI.getSystemUser(), true);
+                if (second.getLanguageId() == context.getLanguageId()) {
+                    idx = 1;
+                }
+            }
+
+            final ContentletSearch contentletSearch = contentletSearches.get(idx);
+            contentlet = this.contentletAPI
+                    .find(contentletSearch.getInode(), this.wuserAPI.getSystemUser(), true);
+
+            this.permissionAPI.checkPermission(contentlet, PermissionLevel.READ, context.getUser());
+        }
+
+        return contentlet;
+    }
+
+    private String buildContentQuery(
+            final Matches matches,
+            final Structure structure,
+            final Field hostField,
+            final UrlMapContext context) {
+
+        final StringBuilder query = new StringBuilder();
+
+        query.append("+structureName:").append(structure.getVelocityVarName())
+             .append(" +deleted:false ");
+
+        if (context.getMode().isAdmin || context.getMode() == PageMode.EDIT_MODE) {
+            query.append("+working:true ");
+        } else {
+            query.append("+live:true ");
+        }
+
+        if (null != hostField && context.getHost() != null) {
+            query.append(this.getHostFilter(context.getHost()));
+        }
+
+        query.append(this.buildFields(structure, matches))
+             .append(" +languageId:").append(context.getLanguageId()).append(" ");
+
+        return query.toString();
+    }
+
+    private boolean shouldLoadPatterns() {
+        String mastRegEx = null;
+
+        try {
+            mastRegEx = CacheLocator.getContentTypeCache().getURLMasterPattern();
+        } catch (DotCacheException e2) {
+            Logger.error(URLMapFilter.class, e2.getMessage(), e2);
+        }
+
+        return mastRegEx == null || patternsCache == null || patternsCache.isEmpty();
+    }
+
+    /**
+     * Builds the list of URL maps and sorts them by the number of slashes in
+     * the URL (highest to lowest). This method is called only when a new URL
+     * map is added, and is marked as <code>synchronized</code> to avoid data
+     * inconsistency.
+     *
+     * @return A <code>String</code> containing a Regex, which contains all the URL maps in the
+     * system.
+     * @throws DotDataException An error occurred when retrieving information from the database.
+     */
+    private synchronized void loadPatterns() throws DotDataException {
+        patternsCache = new TreeSet<>();
```

Comment:
it seems patternsCache is not being a thread-safe collection, it should be mark as a volatile instead

Reply:
That is a great catch - though I don't think we should ever (or rarely) use `volatile` 

Instead, lets use a concurrent set like `ConcurrentHashMap.newKeySet()`

Reply:
we use volatile sometimes when we use a single entity or so that could be change by diff threads *(to avoid the mem-cache scheme), other cases we use thread-safe collections or atomic wrappers, I know there is a few cases were we used, probably have to add doc explaining why we are using it